### ISSUE TITLE
[API Proposal]: Blackhole

### ISSUE THREAD

**Issue by timcassell** [dotnet/runtime#104877]
### Background and motivation

A JIT intrinsic API that does nothing. Similar to Java's [Blackhole](https://javadoc.io/doc/org.openjdk.jmh/jmh-core/1.23/org/openjdk/jmh/infra/Blackhole.html). `Blackhole` is used to prevent dead code elimination and out-of-order instructions generated by a compiler (not by the CPU, so it doesn't emit any barriers). The actual call does nothing, and is stripped out of the assembly.

The purpose of this API is to make it easier/possible to benchmark sensitive operations. Without this API, heavier alternatives are required like calling a method with `NoInlining`, or writing to a field ([see Consumer class in BDN](https://github.com/dotnet/BenchmarkDotNet/blob/01d9b7889e13cf756068985275a3997217e4a102/src/BenchmarkDotNet/Engines/Consumer.cs)), which can spoil the results of the measurement. The cost of this has already been observed (#89940) (also credit to @EgorBo for this idea in that issue).

### API Proposal

```csharp
namespace System.Runtime.CompilerServices;

public static class Blackhole
{
    public void Consume<T>(T value) where T : allows ref struct;
    public void Consume<T>(T* value);
    public void Consume(void* value);
}
```

### API Usage

Old code:

```csharp
public class HalfCast
{
    private readonly Consumer consumer = new();
    
    // 4 operations per invoke averages the cost of the 4 operations
    [Benchmark(OperationsPerInvoke = 4)]
    [Arguments(6.097555E-05f, 12345.0f, 65520.0f, float.NaN)]
    public void SingleToHalfAverage(float arg1, float arg2, float arg3, float arg4)
    {
        // Oh no, the consumer's volatile write interferes with the results!
        consumer.Consume((Half) arg1);
        consumer.Consume((Half) arg2);
        consumer.Consume((Half) arg3);
        consumer.Consume((Half) arg4);
    }
}
```

New code:

```csharp
public class HalfCast
{
    // 4 operations per invoke averages the cost of the 4 operations
    [Benchmark(OperationsPerInvoke = 4)]
    [Arguments(6.097555E-05f, 12345.0f, 65520.0f, float.NaN)]
    public void SingleToHalfAverage(float arg1, float arg2, float arg3, float arg4)
    {
        // Yay, we get accurate results!
        Blackhole.Consume((Half) arg1);
        Blackhole.Consume((Half) arg2);
        Blackhole.Consume((Half) arg3);
        Blackhole.Consume((Half) arg4);
    }
}
```

### Alternative Designs

Pass the value to a method with `NoInlining`, or write to a field. Both of these have some overhead.

**Comment by hamarb123** [comment]
I don't think there'd be any point to a `T*` version if we have a `void*` version (since the conversion does not need to be performed at IL level). It would also be good to have a version that takes `ref readonly T` though (still with the `where T : allows ref struct`) so we can consume references without having to create a ByRef struct just to pass to this API or similar.

**Comment by timcassell** [comment]
> I don't think there'd be any point to a `T*` version if we have a `void*` version (since the conversion does not need to be performed at IL level).

This is true, but the `T*` version is just more ergonomic. Otherwise you'd have to add the cast to your code. Ideally it would just be a single API with an extra `allows pointer` anti-constraint, but that doesn't exist currently.

> It would also be good to have a version that takes `ref T` though (still with the `where T : allows ref struct`) so we can consume references without having to create a ByRef struct just to pass to this API or similar.

That's not necessary.

```cs
ref int target = ref myArray[0];
Blackhole.Consume(target);
```

**Comment by hamarb123** [comment]
> This is true, but the `T*` version is just more ergonomic. Otherwise you'd have to add the cast to your code. Ideally it would just be a single API with an extra `allows pointer` anti-constraint, but that doesn't exist currently.

That's not true - you don't have to explicitly cast to `void*` (it's an implicit cast in c#).

> That's not necessary.
> 
> ```cs
> ref int target = ref myArray[0];
> Blackhole.Consume(target);
> ```

This is not behaviourally or conceptually the same as `Blackhole.Consume(ref target)` - the former explicitly dereferences the byref (which is not a no-op necessarily, for example it can throw in some cases), and is equivalent to (ignoring covariance) `Blackhole.Consume(myArray[0])` from the runtime's perspective. `Blackhole.Consume(ref x)` would consume the by-reference itself, whereas `Blackhole.Consume(x)` just consumes `x`.

**Comment by timcassell** [comment]
You are correct, I updated the proposal.

**Comment by timcassell** [comment]
Should it be `in T` rather than `ref T`? `ref` can go to `in`, but not vice-versa without `Unsafe`.

**Comment by hamarb123** [comment]
From the runtime's perspective, `in T` and `ref T` are the same. This may lead to the issue where you pass something by in and expect the method to treat it as if the ref can't be mutated, but the runtime will most likely not implement it this way (as they're all just `T&` to it). I would add it to the open questions section, and leave it with `ref` (since it's technically more correct).

**Comment by jkotas** [comment]
> prevent dead code elimination and out-of-order instructions generated by a compiler (not by the CPU

Is this actually useful if it does not prevent out-of-order execution done by the CPU?

I think writing the value into a globally visible field to establish data dependency has better fidelity for micro-benchmarking purposes.

**Comment by timcassell** [comment]
> Is this actually useful if it does not prevent out-of-order execution done by the CPU?
> 
> I think writing the value into a globally visible field to establish data dependency has better fidelity for micro-benchmarking purposes.

The idea is for this to be zero cost. As I understand it, barriers that prevent cpu instruction reordering are not free. Anyway, I don't think it's unrealistic for real-world code to do calculations without barriers (storing results in locals instead of fields), so the micro-benchmarks should be fine, I think.

**Comment by jkotas** [comment]
> The idea is for this to be zero cost.

Why does it matter whether this is zero cost? The benchmarking frameworks typically subtract the setup costs from the absolute number. The cost of establishing the global data dependency can be included in the setup costs.

> Also, refs and ref structs cannot be written to globally visible fields

Benchmark.NET can cast ref into a pointer or write the ref struct as raw bytes into a globally visible field to establish the data dependency. Yes, microbenchmarking is hard.

**Comment by timcassell** [comment]
> Why does it matter whether this is zero cost? The benchmarking frameworks typically subtract the setup costs from the absolute number. The cost of establishing the global data dependency can be included in the setup costs.

Because this is meant to be used in the benchmark method itself, not just as part of the benchmark framework. I put an example in the OP. Here's another simple one:

```cs
foreach (var item in collection)
{
    Blackhole.Consume(item);
}
```

> Yes, microbenchmarking is hard.

And this API should make it slightly easier.

**Comment by jkotas** [comment]
What are the chances that the speculative execution done by the processor is going to skew the results a lot if this API is not establishing data dependency? I suspect that the patterns you are proposing would not produce reliable microbenchmarks for modern processors.

> > write the ref struct as raw bytes

> How?

I was originally thinking about taking address of the struct as unmanaged pointer, but it is probably not the best idea. Establishing data dependency for each field of the ref struct is probably a better solution.

**Comment by timcassell** [comment]
> What are the chances that the speculative execution done by the processor is going to skew the results a lot if this API is not establishing data dependency? I suspect that the patterns you are proposing would not produce reliable microbenchmarks for modern processors.

I don't understand how speculative execution impacts on this. In my mind, it's the same as writing the value to a local instead of a field.

Anyway, I'm curious what @EgorBo thinks of having a barrier, since this was originally his idea.

**Comment by julealgon** [comment]
Really dislike this name. Isn't there something more semantic that could be used instead?

**Comment by Windows10CE** [comment]
For further prior art (and an argument for a different signature) I'd like to point to [Rust's std::hint::black_box](https://doc.rust-lang.org/std/hint/fn.black_box.html) which has a similar goal of inhibiting some kinds of optimizations while benchmarking.

I think this API doubling as the identity function is very useful, and allows for usages that normally don't apply, like the ones given in the Rust examples.

(I also happen to think it's a better name, but that's purely subjective)

**Comment by Windows10CE** [comment]
I'd also like to mention you can recreate this behavior without using NoInlining as long as the type you're talking about is supported by Volatile, see (this example)[https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsACgylWgBUAFrGwATVkwDmilQGZKAJnJdyAb2XlHlc9SSUU5ALIAKAJR2HJycAIUFsMABrYIgEbwBWBnIAenJQ8KiY7wsGX18TIIBfQMdi8gBtUQjWPgAZCDBsQVwASSZWDABdUuIXKjcGAwxUsMjo2IGWcgxfUvslIKCAVSZcbAAzTkrq1vbvCAYhiaGMGH480qCANQgwtmEaAHUodphvWDWp0740KfP5hccxAA7ORrrchJw6DA9G8YB8Tmd8k4ikoCkA===]

**Comment by timcassell** [comment]
@Windows10CE That's interesting, I didn't think about the effects of constant optimization in the benchmark. Tbh, I think it's simply better to pass those constants in as arguments to the benchmark method, but I can see the usefulness of that for someone who wants to write their own raw benchmark without using a benchmarking framework.

We can't use global functions in C#, so what should the C# API be? `BlackBox.Identity`?

Also, too bad we don't have `where T : allows pointer` anti-constraint. `T*** Identity<T>(T*** value)`. 😅 

> I'd also like to mention you can recreate this behavior without using NoInlining as long as the type you're talking about is supported by Volatile, see [this example](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsACgylWgBUAFrGwATVkwDmilQGZKAJnJdyAb2XlHlc9SSUU5ALIAKAJR2HJycAIUFsMABrYIgEbwBWBnIAenJQ8KiY7wsGX18TIIBfQMdi8gBtUQjWPgAZCDBsQVwASSZWDABdUuIXKjcGAwxUsMjo2IGWcgxfUvslIKCAVSZcbAAzTkrq1vbvCAYhiaGMGH480qCANQgwtmEaAHUodphvWDWp0740KfP5hccxAA7ORrrchJw6DA9G8YB8Tmd8k4ikoCkA===)

Cool. I bet we could use the new `Volatile.WriteBarrier()` API when it gets implemented to make it work with any type (#98837).

**Comment by Windows10CE** [comment]
> Cool. I bet we could use the new Volatile.WriteBarrier() API when it gets implemented to make it work with any type

I tried with Interlocked.MemoryBarrier() (which is strictly stronger than WriteBarrier) and it didn't work, so I'd assume that won't work either, even that trick is liable to being optimized out at some point I think, its just a trick that works right now for anyone looking for a solution in this issue (and I believe all past versions)

I like BlackBox.Identity with the same three overloads in the original proposal if we're open to this becoming an intrinsic
